// @reciple/client/dist/utils/errorCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUnsupportedModuleErrorOptions = exports.createLoadModuleFailErrorOptions = exports.createCommandRequiredOptionNotFoundErrorOptions = exports.createCommandPreconditionErrorOptions = exports.createCommandHaltErrorOptions = exports.createCommandExecuteErrorOptions = exports.createUnknownCommandTypeErrorOptions = void 0;
const RecipleError_1 = require("../classes/errors/RecipleError");
// TODO: Remove this file
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createUnknownCommandTypeErrorOptions(given) {
    return RecipleError_1.RecipleError.createUnknownCommandTypeErrorOptions(given);
}
exports.createUnknownCommandTypeErrorOptions = createUnknownCommandTypeErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createCommandExecuteErrorOptions(builder, cause) {
    return RecipleError_1.RecipleError.createCommandExecuteErrorOptions(builder, cause);
}
exports.createCommandExecuteErrorOptions = createCommandExecuteErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createCommandHaltErrorOptions(builder, cause) {
    return RecipleError_1.RecipleError.createCommandHaltErrorOptions(builder, cause);
}
exports.createCommandHaltErrorOptions = createCommandHaltErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createCommandPreconditionErrorOptions(builder, cause) {
    return RecipleError_1.RecipleError.createCommandPreconditionErrorOptions(builder, cause);
}
exports.createCommandPreconditionErrorOptions = createCommandPreconditionErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createCommandRequiredOptionNotFoundErrorOptions(optionName, value) {
    return RecipleError_1.RecipleError.createCommandRequiredOptionNotFoundErrorOptions(optionName, value);
}
exports.createCommandRequiredOptionNotFoundErrorOptions = createCommandRequiredOptionNotFoundErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createLoadModuleFailErrorOptions(moduleDisplayName, cause) {
    return RecipleError_1.RecipleError.createLoadModuleFailErrorOptions(moduleDisplayName, cause);
}
exports.createLoadModuleFailErrorOptions = createLoadModuleFailErrorOptions;
/**
 * @deprecated Use `RecipleError` static methods instead
 */
function createUnsupportedModuleErrorOptions(moduleDisplayName) {
    return RecipleError_1.RecipleError.createUnsupportedModuleErrorOptions(moduleDisplayName);
}
exports.createUnsupportedModuleErrorOptions = createUnsupportedModuleErrorOptions;
