// @reciple/client/dist/utils/assertions/commands/assertions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateCommand = exports.validateCommandHalt = exports.validateCommandExecute = exports.validateCommandRequiredMemberPermissions = exports.validateCommandRequiredBotPermissions = exports.validateCommandCooldown = exports.validateCommandType = void 0;
const CommandAssertions_1 = require("../../../classes/assertions/CommandAssertions");
// TODO: Remove this file
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandType(commandType) {
    CommandAssertions_1.CommandAssertions.validateCommand(commandType);
}
exports.validateCommandType = validateCommandType;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandCooldown(cooldown) {
    CommandAssertions_1.CommandAssertions.validateCommandCooldown(cooldown);
}
exports.validateCommandCooldown = validateCommandCooldown;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandRequiredBotPermissions(permissions) {
    CommandAssertions_1.CommandAssertions.validateCommandRequiredBotPermissions(permissions);
}
exports.validateCommandRequiredBotPermissions = validateCommandRequiredBotPermissions;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandRequiredMemberPermissions(permissions) {
    CommandAssertions_1.CommandAssertions.validateCommandRequiredMemberPermissions(permissions);
}
exports.validateCommandRequiredMemberPermissions = validateCommandRequiredMemberPermissions;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandExecute(execute) {
    CommandAssertions_1.CommandAssertions.validateCommandExecute(execute);
}
exports.validateCommandExecute = validateCommandExecute;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommandHalt(halt) {
    CommandAssertions_1.CommandAssertions.validateCommandHalt(halt);
}
exports.validateCommandHalt = validateCommandHalt;
/**
 * @deprecated Use `CommandAssertions` static methods instead
 */
function validateCommand(command) {
    CommandAssertions_1.CommandAssertions.validateCommand(command);
}
exports.validateCommand = validateCommand;
