// @reciple/client/dist/classes/errors/BaseError.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseError = exports.ErrorType = void 0;
const functions_1 = require("../../utils/functions");
const discord_js_1 = require("discord.js");
const fallout_utility_1 = require("fallout-utility");
const strip_ansi_1 = __importDefault(require("strip-ansi"));
/**
 * @deprecated
 */
var ErrorType;
(function (ErrorType) {
    ErrorType[ErrorType["ClientError"] = 1] = "ClientError";
    ErrorType[ErrorType["ModuleError"] = 2] = "ModuleError";
    ErrorType[ErrorType["CommandError"] = 3] = "CommandError";
})(ErrorType || (exports.ErrorType = ErrorType = {}));
/**
 * @deprecated
 */
class BaseError extends Error {
    code;
    constructor(code, errorCodes, ...placeholders) {
        const message = errorCodes[code];
        super(`${fallout_utility_1.kleur.red().bold(code)} ${(0, functions_1.replacePlaceholders)(typeof message === 'function' ? message() : message, ...(0, discord_js_1.normalizeArray)(placeholders))}`);
        this.code = code;
    }
    isCommandError() {
        return this.errorType === ErrorType.CommandError;
    }
    isModuleError() {
        return this.errorType === ErrorType.ModuleError;
    }
    toString() {
        return (0, strip_ansi_1.default)(super.toString());
    }
}
exports.BaseError = BaseError;
