// @reciple/client/dist/classes/errors/ModuleError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleError = exports.ModuleErrorCodes = void 0;
const discord_js_1 = require("discord.js");
const BaseError_1 = require("./BaseError");
const fallout_utility_1 = require("fallout-utility");
// TODO: Remove this class
/**
 * @deprecated
 */
exports.ModuleErrorCodes = {
    /**
     * @deprecated
     */
    LoadModuleFail: () => `Failed to load module ${fallout_utility_1.kleur.blue('%0%')}:\n  ${fallout_utility_1.kleur.red('%1%')}`,
    /**
     * @deprecated
     */
    UnsupportedModule: () => `module ${fallout_utility_1.kleur.blue('%0%')} does not support Reciple ${fallout_utility_1.kleur.green('%1%')}`,
    /**
     * @deprecated
     */
    InvalidScript: () => 'Invalid Reciple module script',
    /**
     * @deprecated
     */
    NoSupportedVersions: () => `Module script is missing the required ${fallout_utility_1.kleur.blue('(versions: string|string[])')} property`,
    /**
     * @deprecated
     */
    InvalidOnStartEvent: () => `Module script's ${fallout_utility_1.kleur.blue('onStart(): boolean|Promise<boolean>')} method is not a valid function`,
    /**
     * @deprecated
     */
    InvalidOnLoadEvent: () => `Module script's ${fallout_utility_1.kleur.blue('(onLoad(): void|Promise<void>)')} method is not a valid function`,
    /**
     * @deprecated
     */
    InvalidOnUnloadEvent: () => `Module script's ${fallout_utility_1.kleur.blue('(onUnload(): void|Promise<void>)')} method is not a valid function`
};
/**
 * @deprecated
 */
class ModuleError extends BaseError_1.BaseError {
    errorType = BaseError_1.ErrorType.ModuleError;
    constructor(code, ...placeholders) {
        super(code, exports.ModuleErrorCodes, ...(0, discord_js_1.normalizeArray)(placeholders));
    }
}
exports.ModuleError = ModuleError;
