// @reciple/client/dist/types/halt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandHaltReason = void 0;
var CommandHaltReason;
(function (CommandHaltReason) {
    CommandHaltReason[CommandHaltReason["Error"] = 1] = "Error";
    CommandHaltReason[CommandHaltReason["Cooldown"] = 2] = "Cooldown";
    CommandHaltReason[CommandHaltReason["InvalidArguments"] = 3] = "InvalidArguments";
    CommandHaltReason[CommandHaltReason["MissingArguments"] = 4] = "MissingArguments";
    CommandHaltReason[CommandHaltReason["ValidateOptionError"] = 5] = "ValidateOptionError";
    CommandHaltReason[CommandHaltReason["MissingMemberPermissions"] = 6] = "MissingMemberPermissions";
    CommandHaltReason[CommandHaltReason["MissingBotPermissions"] = 7] = "MissingBotPermissions";
    CommandHaltReason[CommandHaltReason["NoExecuteHandler"] = 8] = "NoExecuteHandler";
    CommandHaltReason[CommandHaltReason["PreconditionError"] = 9] = "PreconditionError";
})(CommandHaltReason || (exports.CommandHaltReason = CommandHaltReason = {}));
