// @reciple/client/dist/utils/assertions/module/assertions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateModuleScript = exports.validateModuleSlashCommandPrecondition = exports.validateModuleMessageCommandPrecondition = exports.validateModuleContextMenuCommandPrecondition = exports.validateModuleOnUnload = exports.validateModuleOnLoad = exports.validateModuleOnStart = exports.validateModuleScriptVersions = void 0;
const RecipleModuleAssertions_1 = require("../../../classes/assertions/RecipleModuleAssertions");
// TODO: Remove this file
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleScriptVersions(versions) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleScriptVersions(versions);
}
exports.validateModuleScriptVersions = validateModuleScriptVersions;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleOnStart(onStart) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleOnStart(onStart);
}
exports.validateModuleOnStart = validateModuleOnStart;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleOnLoad(onLoad) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleOnLoad(onLoad);
}
exports.validateModuleOnLoad = validateModuleOnLoad;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleOnUnload(onUnload) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleOnUnload(onUnload);
}
exports.validateModuleOnUnload = validateModuleOnUnload;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleContextMenuCommandPrecondition(precondition) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleContextMenuCommandPrecondition(precondition);
}
exports.validateModuleContextMenuCommandPrecondition = validateModuleContextMenuCommandPrecondition;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleMessageCommandPrecondition(precondition) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleMessageCommandPrecondition(precondition);
}
exports.validateModuleMessageCommandPrecondition = validateModuleMessageCommandPrecondition;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleSlashCommandPrecondition(precondition) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleSlashCommandPrecondition(precondition);
}
exports.validateModuleSlashCommandPrecondition = validateModuleSlashCommandPrecondition;
/**
 * @deprecated Use `RecipleModuleAssertions` static methods instead
 */
function validateModuleScript(moduleScript) {
    RecipleModuleAssertions_1.RecipleModuleAssertions.validateModuleScript(moduleScript);
}
exports.validateModuleScript = validateModuleScript;
